var/spack/repos/builtin/packages/amp/package.py

```diff
@@ -3,125 +3,59 @@
-from spack.pkg.builtin.boost import Boost
 
 
-class Amp(CMakePackage):
+class Amp(CMakePackage, CudaPackage, ROCmPackage):
     """The Advanced Multi-Physics (AMP) package.
 
     The Advanced Multi-Physics (AMP) package is an open source parallel
     object-oriented computational framework that is designed with single
     and multi-domain multi-physics applications in mind.
     """
 
-    homepage = "https://bitbucket.org/AdvancedMultiPhysics/amp"
-    hg = homepage
+    homepage = "https://github.com/AdvancedMultiPhysics/AMP"
+    git = "https://github.com/AdvancedMultiPhysics/AMP.git"
 
-    version("develop")
+    maintainers("bobby-philip", "gllongo", "rbberger")
+
+    license("UNKNOWN")
+
+    version("master", branch="master")
+    version("3.1.0", tag="3.1.0", commit="c8a52e6f3124e43ebce944ee3fae8b9a994c4dbe")
 
-    variant("boost", default=True, description="Build with support for Boost")
-    variant("hdf5", default=True, description="Build with support for HDF5")
-    variant("hypre", default=True, description="Build with support for hypre")
-    variant("libmesh", default=True, description="Build with libmesh support")
     variant("mpi", default=True, description="Build with MPI support")
-    variant("netcdf", default=True, description="Build with NetCDF support")
-    variant("petsc", default=True, description="Build with Petsc support")
-    variant("shared", default=True, description="Build shared libraries")
-    variant("silo", default=True, description="Build with support for Silo")
-    variant("sundials", default=True, description="Build with support for Sundials")
-    variant("trilinos", default=True, description="Build with support for Trilinos")
-    variant("zlib", default=True, description="Build with support for zlib")
-
-    # Everything should be compiled position independent (-fpic)
-    depends_on("blas")
-    depends_on("lapack")
-
-    # TODO: replace this with an explicit list of components of Boost,
-    # for instance depends_on('boost +filesystem')
-    # See https://github.com/spack/spack/pull/22303 for reference
-    depends_on(Boost.with_default_variants, when="+boost")
-    depends_on("hdf5", when="+hdf5")
-    depends_on("hypre", when="+hypre")
-    depends_on("libmesh", when="+libmesh")
-    depends_on("netcdf-c", when="+netcdf")
-    depends_on("petsc", when="+petsc")
-    depends_on("silo", when="+silo")
-    depends_on("sundials", when="+sundials")
-    depends_on("trilinos", when="+trilinos")
-    depends_on("zlib-api", when="+zlib")
-
-    # MPI related dependencies
-    depends_on("mpi", when="+mpi")
+    variant("hypre", default=False, description="Build with support for hypre")
+    variant("kokkos", default=False, description="Build with support for Kokkos")
+    variant("openmp", default=False, description="Build with OpenMP support")
+    variant("shared", default=False, description="Build shared libraries")
+
+    depends_on("cmake@3.26.0:")
+    depends_on("tpl-builder+stacktrace")
```

Comment:
@wdconinc The automated build script found that when building `amp+shared`, the link fails with
```py
ld: /home/bkaindl/spack.7/opt/spack/linux-ubuntu22.04-skylake/gcc-13.1.0/stacktrace-0.0.93-2btfwmlhnipxil7eol5zyrrenrv5suxa/lib/libstacktrace.a(StackTraceThreads.cpp.o):
relocation R_X86_64_TPOFF32 against `__tls_guard' can not be used when making a shared object; recompile with -fPIC
```
I assume that this can be fixed by ensuring that `libstacktrace.a` is built with `-fPIC` (+pic) or a `libstacktrace.so` (+shared) is used.

Now, I don't know why one would have to build a non-PIC (non-relocatable) static `libstacktrace.a`, because if this is only for handling error traces, it should never be in the "hot" code path that is performance-critical, that would benefit from not using `-fPIC`. Likewise, I don't know why one would have a big benefit by using a static (`.a`) `libstacktrace` as opposed to a shared one, except for having fewer dynamically linked libraries to link at runtime, as `libstacktrace` should not be performance-relevant.
(no issue, I would just be curious).

Anyway, the default variants of `libstacktrace` are ~pic~shared, thus we get a non-relocatable `libstacktrace.a` by default, and we see this error right away. Otherwise, we'd only see it if for some reason `libstacktrace~pic~shared` is used, and then used for `amp+shared`.
 
```suggestion
    depends_on("tpl-builder+stacktrace")
    depends_on("tpl-builder+stacktrace+shared", when="+shared")
```
Maybe, `tpl-builder` also needs a `+shared` variant in which it depends on `libstackstrace+shared`, I didn't check.

Reply:
Added more constraints for `stacktrace`.